Remora.Discord.Commands/API/ResponseTrackingInteractionAPI.cs

```diff
@@ -0,0 +1,212 @@
+//  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//  GNU Lesser General Public License for more details.
+//
+//  You should have received a copy of the GNU Lesser General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+//
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.DependencyInjection;
+using OneOf;
+using Remora.Discord.API.Abstractions.Objects;
+using Remora.Discord.API.Abstractions.Rest;
+using Remora.Discord.Commands.Contexts;
+using Remora.Discord.Commands.Services;
+using Remora.Rest.Core;
+using Remora.Results;
+
+namespace Remora.Discord.Commands.API;
+
+/// <summary>
+/// Implements a tracking system for whether an interaction has been responded to.
+/// </summary>
+internal class ResponseTrackingInteractionAPI : IDiscordRestInteractionAPI
+{
+    private readonly ContextInjectionService? _contextInjector;
+    private readonly IDiscordRestInteractionAPI _actual;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="ResponseTrackingInteractionAPI"/> class.
+    /// </summary>
+    /// <param name="actual">The actual API instance.</param>
+    /// <param name="provider">The service provider.</param>
+    public ResponseTrackingInteractionAPI
+    (
+        IDiscordRestInteractionAPI actual,
+        IServiceProvider provider
+    )
+    {
+        _actual = actual;
+
+        // TODO: Improve this with a proper scope check once I figure out how to do that
+        try
+        {
+            _contextInjector = provider.GetRequiredService<ContextInjectionService>();
+        }
+        catch (InvalidOperationException)
+        {
+        }
+    }
+
+    /// <inheritdoc />
+    public async Task<Result> CreateInteractionResponseAsync
+    (
+        Snowflake interactionID,
+        string interactionToken,
+        IInteractionResponse response,
+        Optional<IReadOnlyList<OneOf<FileData, IPartialAttachment>>> attachments = default,
+        CancellationToken ct = default
+    )
+    {
+        var createResponse = await _actual.CreateInteractionResponseAsync
+        (
+            interactionID,
+            interactionToken,
+            response,
+            attachments,
+            ct
+        );
+
+        if (_contextInjector?.Context is InteractionContext interactionContext)
+        {
+            // only flip this to true once, and never go back to false
+            interactionContext.HasRespondedToInteraction |= createResponse.IsSuccess;
+        }
+
+        return createResponse;
+    }
+
+    /// <inheritdoc />
+    public Task<Result<IMessage>> GetOriginalInteractionResponseAsync
+    (
+        Snowflake applicationID,
+        string interactionToken,
+        CancellationToken ct = default
+    ) =>
+        _actual.GetOriginalInteractionResponseAsync(applicationID, interactionToken, ct);
```

Comment:
Couldn't an optimization technically be made here to immediately return a `RestResult<NotFoundError>` if a response hasn't been created yet? Would save on API calls.

Reply:
It could not be optimized by simply returning RestResult<NotFoundError> without breaking every other class that is using `IDiscordRestInteractionAPI` as in the `ServiceCollection` the implementation of `IDiscordRestInteractionAPI` is being replaced by `ResponseTrackingInteractionAPI` while wrapping the old registered `IDiscordRestInteractionAPI`.

The calling chain due to the decorator pattern would be as follows: (caching enabled + commands enabled)
```
ResponseTrackingInteractionAPI.GetOriginalInteractionResponseAsync(...) 
calls (delegates to)
  CachingDiscordRestInteractionAPI.GetOriginalInteractionResponseAsync(...) 
  calls (delegates to)
    DiscordRestInteractionAPI .GetOriginalInteractionResponseAsync(...)
```

So any class that is using `IDiscordRestInteractionAPI` would break when you break the delegation chain (by not calling the actual method).